File: QuickSort.java
import java.util.Random;

public class QuickSort {
    private static void swap(int[] arr , int idx1 , int idx2){
        int temp = arr[idx1];
        arr[idx1] = arr[idx2];
        arr[idx2] = temp;
    }


    private static void quickSort(int[] array , int lowIdx , int highIdx){

        if(lowIdx >= highIdx){
            return;
        }

        int pivotIdx = new Random().nextInt(highIdx - lowIdx) + lowIdx;
        int pivot = array[pivotIdx];
        swap(array , pivotIdx , highIdx);


        int leftPointer = lowIdx;
        int rightPointer = highIdx;

        while (leftPointer < rightPointer){
            while(array[leftPointer] <= pivot && leftPointer < rightPointer){
                leftPointer++;
            }

            while(array[rightPointer]>= pivot && leftPointer < rightPointer){
                rightPointer--;
            }

            swap(array , leftPointer , rightPointer);
        }


        swap(array , leftPointer , highIdx);
        quickSort(array , lowIdx , leftPointer -1);
        quickSort(array , leftPointer + 1 , highIdx);
    }

    private static void quickSort(int[] arr){
        quickSort(arr , 0  , arr.length - 1);
    }

    public static void main(String[] args) {
        Random rand = new Random();
        int[] arr = new int[100000];
        int n = arr.length;

        for (int i = 0; i < arr.length; i++) {
            arr[i] = rand.nextInt(1000000);
        }

        quickSort(arr);

        for (int i = 0 ; i < n ; i++){
            System.out.println(arr[i] + " ");
        }
        System.out.println();
    }
}

File: QuickSort.cpp
#include <bits/stdc++.h>
using namespace std;

int partition(vector<int> &arr , int low , int high){
    int pivot = arr[low];
    int i = low;
    int j = high;
    while(i < j){
        while(arr[i] <= pivot && i <= high -1){
            i++;
        }

        while (arr[j] > pivot && j >= low +1){
            j--;
        }

        if(i <j){
            swap(arr[i] , arr[j]);
        }
        
    }

    swap(arr[low] , arr[j]);
    return j;
}

void qs(vector<int> &arr , int low , int high){
    if(low < high){
        int partitionInx = partition(arr , low , high);
        qs(arr , low , partitionInx - 1);
        qs(arr , partitionInx + 1 , high);
    }
}

vector<int> quickSort(vector<int> arr){
    qs(arr , 0 , arr.size() -1);
    return arr;
}

int main(){

    vector<int> Arr = {4 , 1 , 7 , 9 , 3 , 2 , 8};
    Arr = quickSort(Arr);
    for(int i = 0 ; i < Arr.size() ; i++){
        cout<< Arr[i] << " ";
    }
    cout << endl;

    return 0;
}